ca_gazebo/src/traffic_light_plugin.cpp

```diff
@@ -0,0 +1,151 @@
+
+        /// True to use wall time, false to use sim time.
+        public: bool useWallTime;
+
+        /// Subscriber to world info.
+        public: transport::SubscriberPtr infoSub;
+    };
+
+    GazeboTrafficLight::GazeboTrafficLight() : data_ptr(new GazeboTrafficLightPrivate)
+    {}
+
+    GazeboTrafficLight::~GazeboTrafficLight()
+    {}
+
+    void GazeboTrafficLight::Load(rendering::VisualPtr _visual, sdf::ElementPtr _sdf)
+    {
+        // Check for null pointer 
+        if (!_visual || !_sdf)
+        {
+        gzerr << "No visual or SDF element specified. Plugin won't load." <<
+            std::endl;
+        return;
+        }
+        // Get parameters
+        this->data_ptr->visual = _visual;
+        this->red_time_    = common::Time(_sdf-> Get<double>("redTime"));
+        this->yellow_time_ = common::Time(_sdf-> Get<double>("yellowTime"));
+        this->green_time_  = common::Time(_sdf-> Get<double>("greenTime"));
+        // Check for correct time for every color
+        if (this->data_ptr->red_time_ <= 0 || this->data_ptr->yellow_time_ <= 0 || this->data_ptr->green_time_ <= 0)
+        {
+            gzerr << "Time can't be lower than zero." << std::endl;
+            return;
+        }
+        // Connect to the world update signal
+        this->data_ptr->updateConnection = event::Events::ConnectPreRender(std::bind(&GazeboTrafficLight::Update, this));
+        // Subscribe to world statistics to get sim time
+        // Warning: topic ~/pose/local/info is meant for high-bandwidth local
+        // network access. It will kill the system if a remote gzclient tries to
+        // subscribe.
+        // Create our node for communication
+        this->data_ptr->node = transport::NodePtr(new transport::Node());
+        // Initialize the node with the world name
+        this->data_ptr->node->Init();
+        // Listen to Gazebo world_stats topic
+        commandSubscriber = this->data_ptr->node->Subscribe("/gazebo/EmptyWorld/WorldTime_topic", &GazeboTrafficLight::time_update_cb, this);
+        this->next_color_ = "red";
+        this->last_time_ = this->current_time_;
+    }
+
+    void GazeboTrafficLight::Update()
+    {
+        //std::lock_guard<std::mutex> lock(this->data_ptr->mutex);
+        if (!this->data_ptr->visual)
+        {
+            gzerr << "The visual is null." << std::endl;
+            return;
+        }
+
+        common::Time period_time;
+
+        if(this->next_color_ == "red")
+            period_time = this->green_time_;
+        else if(this->next_color_ == "yellow")
+            period_time = this->red_time_;
+        else
+            period_time = this->yellow_time_;
+
+        auto elapsed = this->current_time_ - this->last_time_;
+
+        if(elapsed > period_time)
+        {
+            this->last_time_ = this->current_time_;
+
+            double red = 0;
+            double green = 0;
+            double blue = 0;
+
+            if(this->next_color_ == "red")
```

Comment:
I believe this logic can be done much easier but I don't have a clue now on how to do it.

Reply:
You can use a `std::map` that uses `this->next_color_` as key and its value is the `ignition::math::Color`.